test/unit/StakingProducts/recalculateEffectiveWeight.js

```diff
@@ -0,0 +1,447 @@
+
+    const stakedProduct = await stakingProducts.getProduct(this.poolId, productIdToAdd);
+    expect(stakedProduct.lastEffectiveWeight).to.be.equal(100);
+    expect(stakedProduct.targetWeight).to.be.equal(100);
+
+    const unknownProduct = await stakingProducts.getProduct(this.poolId, productIdToAdd.add(1));
+    expect(unknownProduct.lastEffectiveWeight).to.be.equal(0);
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(100);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(100);
+
+    // recalculating should do nothing
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, [productIdToAdd, productIdToAdd.add(1)]);
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(100);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(100);
+
+    {
+      const unknownProduct = await stakingProducts.getProduct(this.poolId, productIdToAdd.add(1));
+      expect(unknownProduct.lastEffectiveWeight).to.be.equal(0);
+      expect(unknownProduct.targetWeight).to.be.equal(0);
+    }
+  });
+
+  // TODO: implement once https://github.com/NexusMutual/smart-contracts/issues/842 is sorted out
+  it.skip('effective weight should be > target when allocations are greater than capacity', async function () {});
+
+  it('should calculate effective weight properly when decreasing target weight', async function () {
+    const { stakingProducts } = this;
+    const [staker, coverBuyer] = this.accounts.members;
+
+    const amount = parseEther('12345');
+    const coverBuyAmount = amount.mul(8).div(100).mul(2); // 8% of capacity with 2x capacity multiplier
+
+    const productIdToAdd = Zero;
+    await setStakedProducts.call(this, { productIds: [productIdToAdd], targetWeight: 10 });
+
+    // deposit stake
+    await depositTo.call(this, { staker, amount });
+
+    // buy cover
+    await allocateCapacity.call(this, { coverBuyer, amount: coverBuyAmount, productId: productIdToAdd });
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, [productIdToAdd]);
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(10);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(10);
+    {
+      const stakedProduct = await stakingProducts.getProduct(this.poolId, productIdToAdd);
+      expect(stakedProduct.lastEffectiveWeight).to.be.equal(10);
+      expect(stakedProduct.targetWeight).to.be.equal(10);
+    }
+
+    // decrease target weight
+    await setStakedProducts.call(this, { productIds: [productIdToAdd], targetWeight: 5 });
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, [productIdToAdd]);
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(5);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(8);
```

Comment:
This value 8 should be a named constant

it's easier to visualize and understand what's going on

and use computed values as much as possible as well